src/app/modules/submission/IncomingSubmission.class.ts

```diff
@@ -0,0 +1,199 @@
+import { combine, err, ok, Result } from 'neverthrow'
+
+import {
+  FieldIdSet,
+  getLogicUnitPreventingSubmit,
+  getVisibleFieldIds,
+} from '../../../shared/util/logic'
+import {
+  FieldResponse,
+  IFieldSchema,
+  IFormDocument,
+  IPopulatedForm,
+} from '../../../types'
+import { validateField } from '../../utils/field-validation'
+
+import { ProcessingError, ValidateFieldError } from './submission.errors'
+import {
+  ProcessedFieldResponse,
+  ValidatedFieldMap,
+  VerifiableResponseIdSet,
+  VisibleResponseIdSet,
+} from './submission.types'
+
+export abstract class IncomingSubmission {
+  private readonly visibleFieldIds: FieldIdSet
+  private readonly visibleResponseIds: VisibleResponseIdSet
+  private readonly verifiableResponseIds: VerifiableResponseIdSet
+  protected constructor(
+    public readonly responses: FieldResponse[],
+    public readonly form: IPopulatedForm,
+    private fieldMap: ValidatedFieldMap,
+  ) {
```

Comment:
it would be good to add a note here stating that the subclass should be instantiated using the filtered responses from `getModeFilter`, not the raw responses from the frontend. hopefully that's a temporary thing, I suspect more refactors will fall out once we implement `IncomingEmailSubmission`

Reply:
let me take this one step further and encode this requirement as a nominal type. the compiler will crash when instantiating with unfiltered responses 👍